front/scripts/main/models/AddPhotoModel.ts

```
@@ -0,0 +1,156 @@
+				resolve(
+					App.AddPhotoModel.create({
+						title: title,
+						sectionIndex: sectionIndex,
+						photoData: photoData,
+						photoImage: oFREvent.target.result,
+						photoName: photoName,
+						photoExtension: photoExtension
+					})
+				);
+			};
+			oFReader.onerror = function(OFREvent: any): void {
+				reject();
+			};
+		});
+	},
+
+	addToContent: function(uploadedPhotoTitle: any, model: any): Em.RSVP.Promise {
+		var photoWikiText = '\n[[File:' + uploadedPhotoTitle + '|thumb]]\n',
+			data = {
+				action: 'edit',
+				title: model.title,
+				section: model.sectionIndex,
+				format: 'json',
+				prependtext: '',
+				appendtext: '',
+				token: ''
+			};
+		data.appendtext = photoWikiText; //For now, we always add image file at the bottom of section.
+		return new Em.RSVP.Promise((resolve: Function, reject: Function): void => {
+			this.getEditToken(model.title)
+				.then((token: any): void => {
+					data.token = token;
+					Em.$.ajax({
+						url:  Mercury.wiki.basePath + '/api.php',
+						dataType: 'json',
+						method: 'POST',
+						data: data,
+						success: (resp: any): void => {
+							if (resp && resp.edit && resp.edit.result === 'Success') {
+								resolve();
+							} else if (resp && resp.error) {
+								reject(resp.error.code);
+							} else {
+								reject();
+							}
+						},
+						error: (err): void => {
+							reject(err);
+						}
+					});
+				}, (err: any) => {
+					reject(err);
+				});
+		});
+	},
+
+	upload: function(model: any): Em.RSVP.Promise {
+		return new Em.RSVP.Promise((resolve: Function, reject: Function): void => {
+			this.temporaryUpload(model.photoData)
+				.then((addmediatemporary: any): void => {
+					if (addmediatemporary.tempName === undefined) { //we already have the file.
+						resolve({'title': addmediatemporary.title});
+					}
+					var newPhotoTitle = model.photoName.trim().length === 0 ? model.photoData.name : model.photoName.trim() + '.' + model.photoExtension,
```

I asked product requirement for the use case when a user supplies empty image name. We decided that, in such a case, we silently replace empty name with the original file name.